org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
-		}
-
-		this.units[this.currentIndex++] = null;
-		if (this.currentIndex >= this.size)
-			this.currentIndex = 0;
-		if (this.sleepCount >= 1 && ++this.sleepCount > 4) {
-			notify(); // wake up processing thread to add next unit but only after removing some elements first
-			yield = this.sleepCount > 8;
-		}
+				if (!elements.isEmpty()) {
+					if (next !=null) {
+						// defer any stop signal until all CompilationUnitDeclaration read
+						this.lastSignal= next;
+					}
+					return elements;
+				}
+			} catch (InterruptedException interrupt) {
+				throw new RuntimeException(interrupt);
+			}
+			if (next instanceof Error error) {
+				throw error;
+			}
+			if (next instanceof RuntimeException runtimeException) {
+				throw runtimeException;
+			}
+			if (next == STOP_SIGNAL) {
+				return null;
+			}
+			throw new IllegalStateException(String.valueOf(next));
+		} while (true);
 	}
-	if (yield)
-		Thread.yield();
-	return next;
-}
 
-private void compile() {
-	int unitIndex = this.startingIndex;
-	synchronized (this) { // wait until processingTask is assigned
-		@SuppressWarnings("unused")
-		Future<?> p = this.processingTask;
-	}
-	boolean noAnnotations = this.compiler.annotationProcessorManager == null;
-	while (this.processingTask != null) {
-		this.unitToProcess = null;
-		int index = -1;
-		boolean cleanup = noAnnotations || this.compiler.shouldCleanup(unitIndex);
+	private void processing() {
 		try {
-			synchronized (this) {
-				if (this.processingTask == null) return;
+			int unitIndex = this.startingIndex;
+			boolean noAnnotations = this.compiler.annotationProcessorManager == null;
+			while (true) {
+				CompilationUnitDeclaration unitToProcess;
+				int index = -1;
+				boolean cleanup = noAnnotations || this.compiler.shouldCleanup(unitIndex);
+				try {
+					synchronized (this) {
+						unitToProcess = this.compiler.getUnitToProcess(unitIndex);
+						this.lastUnitToProcess = unitToProcess;
+						if (unitToProcess == null) {
+							break;
+						}
+						index = unitIndex++;
+						if (unitToProcess.compilationResult.hasBeenAccepted) {
+							continue;
+						}
+					}
 
-				this.unitToProcess = this.compiler.getUnitToProcess(unitIndex);
-				if (this.unitToProcess == null) {
-					this.processingTask = null;
-					return;
-				}
-				index = unitIndex++;
-				if (this.unitToProcess.compilationResult.hasBeenAccepted)
-					continue;
-			}
+					try {
+						this.compiler.reportProgress(Messages.bind(Messages.compilation_processing,
+								new String(unitToProcess.getFileName())));
+						if (this.compiler.options.verbose)
+							this.compiler.out.println(Messages.bind(Messages.compilation_process,
+									new String[] { String.valueOf(index + 1), String.valueOf(this.compiler.totalUnits),
+											new String(unitToProcess.getFileName()) }));
+						try {
+							this.compiler.process(unitToProcess, index);
+						} catch (AbortCompilation abortCompilation) {
+							unitToProcess.cleanUp();
```

Comment:
`unitToProcess.cleanUp();` seem to be added here - why? It is done in finally below.

Reply:
good question, i forgot why. i will remove it